### PR TITLE
Properly handle renego for initiator-side stream changes

### PR DESCRIPTION
**Pull request by xdumaine**
@fippo @legastero @latentflip This should resolve this bug that I opened last year. When I opened that, I didn't fully understand jsep and offer/answer. This basically allows on the fly stream changes for 1:1 jingle sessions. It's never been a problem before since most users of this who are doing stream changes are connected an SFU (jitsi) and are never the initiator. In my case, I'm allowing stream changes (i.e., adding screen, switching device stream) during 1:1s, which fails because the assumption is always made that when making or processing stream changes, you're the answerer.

tldr: check `isInitiator` and flip the offer/answer to make stream renegos work for initiator side.

 https://github.com/otalk/jingle.js/issues/24

### CODE REVIEW

**Comment by fippo** [comment]
I think this code has an old assumption that you need to preserve the offer/answer role. I don't think that is required anymore even though it is not a bad idea.

There are some other issues here such as the lack of a Q. If you have incoming operations at the wrong time things will break.

I'll see if I can find some time to properly review this over the weekend

**Comment by xdumaine** [comment]
Chrome chokes with wrong actpass if you don't preserve the role.

**Comment by xdumaine** [comment]
Didn't link to the original bug - https://github.com/otalk/jingle.js/issues/24

**Comment by fippo** [comment]
that sounds like there is something wrong in the mangling code somewhere. Chrome is perfectly happy to renegotiate with the answerer becoming the new offerer (at transaction-level, not at session-level)

**Comment by latentflip** [comment]
Im not totally following this ad I'm still getting my head around jingle. But to be clear this is causing issues for me because I'm trying to add two streams from a single peer, and doing so sequentially with add stream causes the issue.

Philip Roberts

**Comment by xdumaine** [comment]
@latentflip this is tangentially related to your issue. I think we need this as well as a queuing fix for yours.

**Comment by latentflip** [comment]
Derp sorry I was reading via email and thought this was my issue we were talking about :)

Philip Roberts

**Comment by xdumaine** [comment]
image

**Comment by xdumaine** [comment]
oops..

**Comment by konradkierus** [comment]
@xdumaine Thanks for this. A few inputs from my side:

- You've accidentally removed addStream in switchStream method:
https://github.com/otalk/jingle-media-session/pull/24/files#diff-168726dbe96b3ce427e7fedce31bb0bcL333
- I've added updating 'senders' property to send correct offers/answers with sendrecv/sendonly/recvonly for initiator side during swtichStream like here: https://github.com/konradkierus/jingle-media-session/commit/c61083afe34cdbe72e6f2254cbe83263036c61e3#diff-168726dbe96b3ce427e7fedce31bb0bcR397
and for both sides during adding source like here: https://github.com/konradkierus/jingle-media-session/blob/c61083afe34cdbe72e6f2254cbe83263036c61e3/index.js#L612
This isn't particularly related with your changes but they are required to update streams in [react-native-webrtc](https://github.com/oney/react-native-webrtc) which uses chromium native webrtc code under the hood.
I didn't want to create separate pull request for this as I didn't test this in SFU/MCU architecture, only P2P but it may be helpful for someone else when he will be doing integration on mobile apps.
